src/views/tutorial-view/server.ts

```diff
@@ -36,12 +36,23 @@ export interface TutorialPageProps {
 export async function getTutorialPageProps(
   product: LearnProductData,
   slug: [string, string]
-): Promise<{ props: TutorialPageProps }> {
+): Promise<TutorialPageProps | null> {
   const { collection, tutorialReference } = await getCurrentCollectionTutorial(
     product.slug as ProductOption,
     slug
   )
+
+  // the tutorial doesn't exist in collection - return 404
+  if (tutorialReference.dbSlug === null || collection.data === null) {
+    return null
+  }
+
   const fullTutorialData = await getTutorial(tutorialReference.dbSlug)
+  // double guard if tutorial doesn't exist after call - return 404
+  if (fullTutorialData === null) {
```

Comment:
This is an extra safety guard, in most cases if the tutorial doesn't exist, we wouldn't get to this call. But I think this is still an important safety guard.